### ISSUE TITLE
RFC Decoupling the version of pyodide-build and Pyodide

### ISSUE THREAD

**Issue by ryanking13** [pyodide/pyodide#4557]
## Proposal Overview

This issue proposes to decouple the versioning of the pyodide-build and the Pyodide distribution, so that we can release new versions of the pyodide-build without having to release a new version of the Pyodide distribution.

1. This will allow us to release new versions of the pyodide-build more frequently, and to make it easier to maintain the pyodide-build.
2. This aligns with the goal of the out-of-tree build, as it is becoming more and more impossible to maintain the full packages sets in the Pyodide distribution, and we need to delegate the maintenance of the packages and recipes to the community.

As this has been discussed a few times in the past since we started supporting out-of-tree builds, I believe we have consensus on this proposal. So this issue mostly focuses on a way to implement this.

## Current Challenges

Currently, the version of the pyodide-build is highly coupled with the version of the Pyodide distribution.

- all build envs are defined in the Makefile.envs file, which is part of the Pyodide distribution.
- pyodide-build’s source code resides in the same repository as Pyodide.
- certain test dependencies for Pyodide are defined within pyodide-build.
- etc

### Support for multiple xbuildenv versions in pyodide-build

Currently, the pyodide-build only supports to install a corresponding version of the xbuildenv. We need to make it possible to install multiple versions of the Pyodide xbuildenv in a single pyodide-build version.

I think the expected form is for pyodide-build to behave like a _version manager, like emsdk, pyenv, or nvm_. This involves introducing new subcommands for listing, installing, and using various xbuildenv versions:

- `pyodide xbuildenv list`: List all possible xbuildenv versions.
- `pyodide xbuildenv install <version>`: Install a specific version of the xbuildenv.
  - This command already exists but for now it only supports to install a corresponding version of the xbuildenv.
- `pyodide xbuildenv use <version>`: Use a specific version of the xbuildenv.

To support this, we will need to maintain a metadata file that contains the information of the xbuildenv versions. This is similar to how [emsdk manages the versions of the emscripten SDK](https://github.com/emscripten-core/emsdk/blob/main/emscripten-releases-tags.json). This metadata file should be hosted on the web separately from pyodide-build so that we can update it when a new version of Pyodide (xbuildenv) is released.

```json
{
    "versions": [
        {
            "version": "0.26.0a1",
            "url": "<xbuildenv tarball url>",
            "sha256": "<sha256 of the tarball>",
            "python_version": "3.12.1",
            "emscripten_version": "...",
            "minimum_pyodide_build_version": "...",
        },
        {
            "version": "0.25.0",
            "url": "<xbuildenv tarball url>",
            "sha256": "<sha256 of the tarball>",
            "python_version": "3.11.3",
            "emscripten_version": "...",
            "minimum_pyodide_build_version": "...",
        },
        {
        
        }
    ]
}
```

This metadata file should include the **pyodide version**, **url to the xbuildenv tarball**. Additionally, having the **python version** in the metadata file would be helpful so that we can check the host Python version is compatible with the xbuildenv version before installing the xbuildenv. Also, having the **minimum pyodide-build version** would be helpful to check the compatibility of the xbuildenv version with the pyodide-build version.

We also need to update the `.pyodide-xbuildenv` directory structure to support multiple versions of the xbuildenv. I suggest the following directory structure, where the `xbuildenv` directory is a symlink to the selected xbuildenv version.

```
.pyodide-xbuildenv-<pyodide-build-version>/
    xbuildenv/ ==> symlink to the selected xbuildenv version
    0.25.0/
        ...
    0.25.1/
        ...
    ...
```

### Decoupling build variables used by pyodide-build from Makefile.envs

We are storing most of the build envs in the Makefile.envs file. Some of them are used in building the Pyodide distribution, while some of them are only used in the pyodide-build. For instance, flags like `HOSTSITEPACKAGES`, `SIDE_MODULE_LDFLAGS`, `RUST_TOOLCHAIN` are only used in the pyodide-build. We need to remove the variables used by the pyodide-build only from the Makefile.envs file so that these variables can be managed separately from the Pyodide distribution.

Some tricky parts of this is that we need to keep some flags in sync between the Pyodide distribution and the pyodide-build. For instance, CFLAGS and LDFLAGS of main module (pyodide) and the side module (packages) should be in sync. We need to carefully check which variables can be updated separately from Pyodide.

### Remove features that are not related to the package build from the pyodide-build

Currently, the pyodide-build contains some features that are not related to the package build:

- `pyodide create-zipfile`: This command is used to create a stdlib zip file from the Python standard library. This is not related to the package build, but it is required to build the Pyodide distribution.

- (Optional) `pyodide py-compile`: This command is used to compile the .py files to .pyc files.

`pyodide create-zipfile` is highly coupled with the Pyodide distribution, so we need to remove this command from the pyodide-build.

### Reevaluating the `pyodide xbuildenv create` command

Given the decoupling, probably `pyodide xbuildenv create` command should not exist in the pyodide-build. This is the most tricky part of this proposal, as for now, quite a lot of xbuildenv creation logic relies on the pyodide-build. IMO, we can keep the `pyodide xbuildenv create` command in the pyodide-build for now. But in the long term, we need to move this out of the pyodide-build and make a separate tool for creating the xbuildenv. This tool may still rely on the pyodide-build, but it should be a separate tool as it relies more on the Pyodide distribution than the pyodide-build.

### (Optional) Move the pyodide-build source code to a separate repository

This is not a must-have, but it would be great to move the pyodide-build source code to a separate repository, if we make the above changes successfully.

## Discussion Points and Concerns

- Making changes in xbuildenv and pyodide-build at the same time will be a bit tricky after the decoupling.
- pyodide-build (pywasmcross) should be compatible to multiple emscripten versions too. Probably we need to add a minimum emscripten version supported by the pyodide-build
- we will need more robust testing for the out-of-tree build to make sure that the pyodide-build is compatible with the multiple versions of xbuildenvs

WDYT @rth @hoodmane?

**Comment by hoodmane** [comment]
Sounds great! Thanks so much for working on this @ryanking13, it's quite an involved project.

**Comment by hoodmane** [comment]
Currently the behavior of xbuildenv is pretty dumb it downloads a new version in each directory. What do you think about sticking the versions in xdg-cache and making a symlink?

**Comment by ryanking13** [comment]
> Currently the behavior of xbuildenv is pretty dumb it downloads a new version in each directory. What do you think about sticking the versions in xdg-cache and making a symlink?

Yes, I think we can use a common directory like user home directory to install the xbuildenv (and probably make is possible for users override it by setting an env variable)

**Comment by rth** [comment]
Thanks for doing this analysis @ryanking13 ! I agree with all of the above points.

**Comment by ryanking13** [comment]
Closing as completed.
